velox/connectors/hive/iceberg/tests/TransformE2ETest.cpp

```diff
@@ -0,0 +1,646 @@
+      {4, TransformType::kIdentity, std::nullopt}, // c_decimal.
+      {5, TransformType::kIdentity, std::nullopt}, // c_varbinary.
+      {3, TransformType::kIdentity, std::nullopt}, // c_varchar2.
+  };
+
+  auto outputDirectory = writeBatchesWithTransforms(
+      {rowVector}, asRowType(rowVector->type()), partitionTransforms);
+
+  const auto& actualPartitionNames =
+      listDirectoriesRecursively(outputDirectory->getPath());
+
+  // Build expected partition folder names.
+  std::vector<std::string> expectedPartitionNames = {
+      "c_int=42",
+      "c_bigint=9876543210",
+      "c_varchar=test+string",
+      "c_decimal=1234567.890",
+      "c_varbinary=YmluYXJ5AGRhdGEBAgM%3D",
+      "c_varchar2=null",
+  };
+
+  ASSERT_EQ(actualPartitionNames, expectedPartitionNames)
+      << "Partition folder names do not match expected values";
+}
+
+TEST_F(TransformE2ETest, bucket) {
+  constexpr int32_t numBuckets = 4;
+  auto rowType = ROW({"c_varchar"}, {VARCHAR()});
+  auto vectors =
+      createTestData(rowType, kDefaultNumBatches, kDefaultRowsPerBatch);
+
+  auto outputDirectory = writeBatchesWithTransforms(
+      vectors, rowType, {{0, TransformType::kBucket, numBuckets}});
+
+  const auto& partitionDirs = firstLevelDirectories(outputDirectory->getPath());
+  ASSERT_LE(partitionDirs.size(), numBuckets);
+
+  int32_t totalRowsInPartitions = 0;
+  for (const auto& dir : partitionDirs) {
+    auto splits = createSplitsForDirectory(dir);
+    auto countPlan = PlanBuilder().tableScan(rowType).planNode();
+    auto partitionRowCount =
+        AssertQueryBuilder(countPlan).splits(splits).runWithoutResults();
+
+    totalRowsInPartitions += partitionRowCount;
+    ASSERT_GT(partitionRowCount, 0);
+  }
+
+  ASSERT_EQ(totalRowsInPartitions, kDefaultNumBatches * kDefaultRowsPerBatch);
+
+  std::unordered_map<std::string, int32_t> valueToExpectedBucket;
+
+  for (const auto& dir : partitionDirs) {
+    const auto dirName = std::filesystem::path(dir).filename().string();
+    const auto& [k, v] = parsePartitionDirName(dirName);
+    const int32_t expectedBucket = std::stoi(v);
+
+    auto dataPlan =
+        PlanBuilder().tableScan(rowType).project({"c_varchar"}).planNode();
+    const auto& dataResult = AssertQueryBuilder(dataPlan)
+                                 .splits(createSplitsForDirectory(dir))
+                                 .copyResults(opPool_.get());
+
+    auto varcharColumn = dataResult->childAt(0)->asFlatVector<StringView>();
+    for (auto i = 0; i < dataResult->size(); i++) {
+      StringView value = varcharColumn->valueAt(i);
+
+      auto computedBucket = computeBucketHash(value, numBuckets);
+      ASSERT_EQ(computedBucket, expectedBucket)
+          << "Value '" << value << "' in partition " << dirName
+          << " has computed bucket " << computedBucket
+          << " but expected bucket " << expectedBucket;
+
+      // Track the bucket for this value to verify consistency.
```

Comment:
Why do we need this validation if we have already asserted that bucket value matches expectedBucket?